### PR TITLE
Various chat input fixes

### LINKED ISSUE #2049: Edit message bugs + line jump
When you swipe to edit message, the old text is not displayed in editor.

Also, when re-editing the message, the initial message is shown. See video below.

Lastly, jump to new line to make a paragraph does not show properly. 
Eg. Typing this:
Line 1
Line 2 

Line 4

Is shown like this:

Line 1
Line 2 
Line 4

https://github.com/user-attachments/assets/08106dd9-eac7-4f92-8abf-431f9efcb0db

### PR DESCRIPTION
**Pull request by gtalha07**
### Fixes the following issues by order of commit:
- Fix for cursor jumping at end of text length when try selection editing. bf623b1 (Discovered in [#2040](https://github.com/acterglobal/a3/issues/2040#issuecomment-2275657319)).
- Optimize suggestion container height and require at least one alphabet to trigger suggestions. cda65d4
- Fix #2049. Commits -> 5fef663 && 8687e81  | ```flutter_matrix_html``` [fix](https://github.com/acterglobal/flutter_matrix_html/commit/23de151fefc012b25649321742cc2c3c2dca6b76)
- Fix meta issue [#456](https://github.com/acterglobal/a3-meta/issues/456). b67835d

### CODE REVIEW

File: app/lib/features/chat/widgets/pill_builder.dart [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import 'package:acter/common/toolkit/buttons/room_chip.dart';
+import 'package:acter/common/toolkit/buttons/user_chip.dart';
+import 'package:flutter/material.dart';
+
+Widget pillBuilder({
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/pill_builder.dart:5]: let's make this an actual widget please.

File: app/lib/features/chat/utils.dart [hunk 2]
```diff
@@ -8,239 +10,292 @@ import 'package:flutter_riverpod/flutter_riverpod.dart';
...
-    case 'm.sticker':
-    case 'm.room.redaction':
-    case 'm.room.encrypted':
-      // supported
-      return true;
-    case 'm.room.message':
-      // special case of supporting locations
-      return message.metadata?['msgType'] == 'm.location';
-  }
+class ChatUtils {
```
Comment (gnunicorn) [inline app/lib/features/chat/utils.dart:15]: why wrap all that into a class with static functions? What's the benefit of that?

File: app/lib/features/chat/utils.dart
Comment (gnunicorn) [inline app/lib/features/chat/utils.dart:1]: I assume that aside from wrapping these into a class nothing changed - the diff is impossible to read - and will skim over the changed.

File: app/lib/features/chat/widgets/bubble_builder.dart [hunk 3]
```diff
@@ -21,7 +21,9 @@ import 'package:swipe_to/swipe_to.dart';
 
 final _log = Logger('a3::chat::bubble_builder');
 
-class BubbleBuilder extends ConsumerWidget {
+// keep this widget stateful consumer
+// otherwise doesn't hold correct message state
+class BubbleBuilder extends ConsumerStatefulWidget {
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/bubble_builder.dart:26]: but ... does that mean our internal provider _doesn't_ have the correct state? Does that mean that once I leave the room and re-enter it, it will show the wrong state? Why would we make the _widget_ be the single source of truth about state if the we used _riverpod_ for state management?

File: app/lib/features/chat/widgets/custom_input.dart [hunk 4]
```diff
@@ -685,10 +681,21 @@ class _TextInputWidgetConsumerState extends ConsumerState<_TextInputWidget> {
...
   void initState() {
     super.initState();
-    WidgetsBinding.instance.addPostFrameCallback((_) {
-      widget.controller.text = ref.read(
-        chatInputProvider.select((value) => value.message),
-      );
+    ref.listenManual(
+        chatInputProvider.select((state) => state.selectedMessageState),
+        (prev, next) {
+      if (next == SelectedMessageState.edit) {
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/custom_input.dart:687]: there was a comment here explaining what is happening. Why did you remove that?

File: app/lib/features/chat/widgets/mention_profile_builder.dart [hunk 5]
```diff
@@ -54,55 +54,52 @@ class MentionProfileBuilder extends ConsumerWidget {
...
-                  onTap: () {
-                    final autocomplete = MultiTriggerAutocomplete.of(context);
-                    ref
-                        .read(chatInputProvider.notifier)
-                        .addMention(displayName, id);
-                    return autocomplete.acceptAutocompleteOption(
-                      displayName.isNotEmpty ? displayName : id.substring(1),
+        return Card(
+          margin: const EdgeInsets.symmetric(horizontal: 8),
+          elevation: 2,
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/mention_profile_builder.dart:59]: what's the reasoning for removing the sized box?

File: app/lib/features/chat/widgets/text_message_builder.dart [hunk 6]
```diff
@@ -131,43 +130,48 @@ class _TextWidget extends ConsumerWidget {
...
-                            color: isNotice
-                                ? Theme.of(context)
-                                    .colorScheme
-                                    .onSurface
-                                    .withOpacity(0.5)
-                                : null,
-                          ),
-                  maxLines: isReply ? 3 : null,
+        enlargeEmoji
+            ? Text(
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/text_message_builder.dart:137]: so... you removed `messageWidth` being used, but it is still part of the widget's _required_ parameters ... why?

and doesn't this means messages now occupy the entire width? I don't think we want that..

File: app/lib/features/member/dialogs/show_member_info_drawer.dart [hunk 7]
```diff
@@ -10,18 +11,30 @@ Future<void> showMemberInfoDrawer({
...
+          builder: (context) => AlertDialog(
+            content: MemberInfoDrawer(
+              key: key,
+              roomId: roomId,
+              memberId: memberId,
+              isShowActions: isShowActions,
+            ),
+          ),
+        )
+      : await showModalBottomSheet(
```
Comment (gnunicorn) [inline app/lib/features/member/dialogs/show_member_info_drawer.dart:26]: what is the reasoning behind this? why didn't say we want user-dialogs to be center dialogs on larger screens ... is that fixing anything?

**Review by gnunicorn — CHANGES_REQUESTED** [review]
There are several changes in this PR that are hard to follow and while the commits are linked to specific fixes the _why_ of the specific changes isn't explained: e.g. Why are we switching from the provider-based messaging to having it on the text controller again? What does it fix and what about the potential fallout of this change? Why put the helper functions into a class? To name a few.

It is hard to judge whether these changes are necessary, actually fix the thing and moreover don't have other side-effects.

This is a particuarly complex piece of UI-code, there are many things interconnected and interwoven and a lot can go wrong when changing things. Unfortunately we lack tests for this and  thus every change needs to be carefully considered and all its side-effects analyzed and looked at. 

For the future may I recommend that if a single commit fixes a single thing, make that single commit its own PR. That way it is easier to review and discuss potential problems, than in a collection of things. And if things built upon each other, that's fine to, just put a reference in the second PR saying that it is built on top of the other. But right now, I can't sign off on this as I understand what the reasons are for most changes, it's hard to understand what its implications are and whether they have been addressed.

File: app/lib/features/chat/utils.dart (hunk 2 above)
Comment (gtalha07) [reply app/lib/features/chat/utils.dart:15]: This is just an attempt to improve methods identification. While I see we have many utilities now for each features, this is becoming confusing as most of them follow the same functional naming but hard to judge where it's actually intended and used. Having class wrapper under them is I think a good way for recognition and readability.

File: app/lib/features/chat/utils.dart
Comment (gtalha07) [reply app/lib/features/chat/utils.dart:1]: Yes, nothing changed as such aside from it.

File: app/lib/features/chat/widgets/custom_input.dart (hunk 4 above)
Comment (gtalha07) [reply app/lib/features/chat/widgets/custom_input.dart:687]: I've skimmed it down to only listen ```SelectedMessageState``` events because keeping ```SelectedMessage``` is causing inconsistent frequent re-renders which resets the controller which is what we've experienced in past to avoid mixing text editing controllers with riverpod. I will bring back the comment with updated action it does.

File: app/lib/features/chat/widgets/mention_profile_builder.dart (hunk 5 above)
Comment (gtalha07) [reply app/lib/features/chat/widgets/mention_profile_builder.dart:59]: The ```SizedBox``` was acting as constraint to limit the height of suggestions list or it was thought to be. But it came to discovery that it is ineffective of resizing layout of ```Overlay``` (Portal) of internal mentions plugin which seems to have its mechanism of sizing using ```widthFactor``` and ```heightFactor```. In cda65d4, you can see I've extended plugin to include ```heightFactor``` param to adjust suggestion height.

File: app/lib/features/member/dialogs/show_member_info_drawer.dart (hunk 7 above)
Comment (gtalha07) [reply app/lib/features/member/dialogs/show_member_info_drawer.dart:26]: I thought of it as UX-improvement as bottomsheet aren't usually well suited on larger screens, especially on large screenWidth having split views...

File: app/lib/features/chat/widgets/text_message_builder.dart (hunk 6 above)
Comment (gtalha07) [reply app/lib/features/chat/widgets/text_message_builder.dart:137]: Oh I think this was unintentional. Apologies! I will undo it.

File: app/lib/features/chat/widgets/bubble_builder.dart (hunk 3 above)
Comment (gnunicorn) [reply app/lib/features/chat/widgets/bubble_builder.dart:26]: but the question is why? Why would that come about?

File: app/lib/features/chat/providers/notifiers/chat_input_notifier.dart [hunk 8]
```diff
@@ -23,45 +21,10 @@ class ChatInputNotifier extends StateNotifier<ChatInputState> {
     );
   }
 
-  void updateMessage(String value) {
-    state = state.copyWith(message: value);
-  }
-
   void setEditMessage(Message message) {
```
[the remark this answers is not in the document]
Comment (gnunicorn) [reply app/lib/features/chat/providers/notifiers/chat_input_notifier.dart:24]: but... if we _removed_ it from the provider then we won't be holding the proper state anymore on re-entry now, do we? Like the messages are gone that we typed ... that is clearly a regression to what we had before, no?

[the remark this answers is not in the document]
Comment (gtalha07) [reply app/lib/features/chat/providers/notifiers/chat_input_notifier.dart:24]: They aren't being added anyway as we aren't holding state of typed messages in ```ActerTriggerTextEditingController```. This change is only affecting the part ```onChanged:``` of text field where it's re-evaluating every time as message is composed, which sounds bad as the controller part resets in result of riverpod updates and widget re-rendering. I think it's much more safe to handle this differently using ```ComposerDraft``` and using that to update controller. It's already in the [works](https://github.com/acterglobal/a3/pull/2056)
